classes/property_list.php

```diff
@@ -0,0 +1,295 @@
+        if (is_array($result)) {
+            $result = new CFDictionary($result); // Convert back to CFDictionary so serialization is recursive.
+            $result = $result->toArray(); // Serialize.
+        }
+        return $result;
+    }
+
+    /**
+     * Update the value of any element with matching key.
+     *
+     * @param string $key Key of element to update.
+     * @param mixed $value  Value to update element with.
+     */
+    public function update_element_value(string $key, $value) {
+        $this->plist_map( function($elvalue, $elkey, $parent) use ($key, $value) {
+            // Set new value.
+            if ($key === $elkey) {
+                $parent->get($elkey)->setValue($value);
+            }
+        }, $this->cfpropertylist->getValue());
+    }
+
+    /**
+     * Update the array of any dict or array element with matching key.
+     *
+     * Will replace array.
+     *
+     * @param string $key Key of element to update.
+     * @param array $value  Array to update element with.
+     */
+    public function update_element_array(string $key, array $value) {
+        $this->plist_map( function($elvalue, $elkey, $parent) use ($key, $value) {
+            if ($key === $elkey) {
+                $element = $parent->get($elkey);
+                // Replace existing element with new element and array but same key.
+                if ($element instanceof CFDictionary) {
+                    $parent->del($elkey);
+                    $parent->add($elkey, new CFDictionary($value));
+                } else if ($element instanceof CFArray) {
+                    $parent->del($elkey);
+                    $parent->add($elkey, new CFArray($value));
+                }
+            }
+        }, $this->cfpropertylist->getValue());
+    }
+
+    /**
+     * Delete any element with a matching key.
+     *
+     * @param string $key Key of element to delete.
+     */
+    public function delete_element(string $key) {
+        $this->plist_map( function($elvalue, $elkey, $parent) use ($key) {
+            // Convert date to iso 8601 if date object.
+            if ($key === $elkey) {
+                $parent->del($key);
+            }
+        }, $this->cfpropertylist->getValue());
+    }
+
+    /**
+     * Convert the PList to XML.
+     *
+     * @return string XML ready for creating an XML file.
+     */
+    public function to_xml() : string {
+        return $this->cfpropertylist->toXML();
+    }
+
+    /**
+     * Return a JSON representation of the PList. The JSON is constructed to be used to generate a SEB Config Key.
+     *
+     * See the developer documention for SEB for more information on the requirements on generating a SEB Config Key.
+     * https://safeexambrowser.org/developer/seb-config-key.html
+     *
+     * @return string A json encoded string.
+     */
+    public function to_json() : string {
+        // Create a clone of the PList, so main list isn't mutated.
+        $jsonplist = new CFPropertyList();
```

Comment:
maybe abstract this as we replicate this code in a couple of places?

Reply:
Only in the constructor and here, and they are handled differently in both places now I've removed the static create method (To be uploaded soon).